src/macros.jl

```
@@ -313,8 +316,29 @@ function parse_system(exprs)
     nsets = length(constraints)
     nsets > 3 && throw(ArgumentError("cannot parse $nsets set constraints"))
 
-    return dynamic_equation, AT, constraints,
-           state_var, input_var, noise_var, dimension
+    # error handling for variable names
+    got_state_var = state_var != nothing
+    got_input_var = input_var != nothing
+    got_noise_var = noise_var != nothing
+    if (got_state_var && got_input_var) && (state_var == input_var)
+         throw(ArgumentError("state and input variables have the same name `$(state_var)`"))
+    elseif (got_state_var && got_noise_var) && (state_var == noise_var)
+         throw(ArgumentError("state and noise variables have the same name `$(state_var)`"))
+    elseif (got_input_var && got_noise_var) && (input_var == noise_var)
+         throw(ArgumentError("input and noise variables have the same name `$(input_var)`"))
+    end
+
+    # assign default values
+    if !got_state_var
+        state_var = default_state_var
+    end
+    if !got_input_var
+        input_var = default_input_var
+    end
+    if !got_noise_var
+        noise_var = default_noise_var
+    end
```

I think you want to make sure that the default is different from the assigned names. Otherwise you run into the same problem as before.

---

Not sure, why? If a default is the same as an assigned name, that is not a problem.

---

You can reach a state below this block where `state_var == input_var`. Why is this not a problem?

---

We only want to throw an error if we changed the default values which is checked e.g by `got_state_var`

---

The code *only* checks for a clash between defined variable names. Example: `state_var: y, input_var: y`.
But the code *does not* check for a clash with the names that will be used later (after assigning the default values). Example: `state_var: nothing, input_var: x`. After assigning the default, you get `state_var: x, input_var: x`. The fact that this does not crash in practice is that `got_state_var` is invariably `true` because that is read from the left-hand side of the equation. So I suggest to remove this variable as it is confusing.
Here is an example that shows what I originally meant (using `input: u` and `noise: nothing`):
```julia
julia> @system(x⁺ = Ax + Bu + Du, input: u)
ERROR: ArgumentError: the entry (:A, :B, :B) does not match a `MathematicalSystems.jl` structure
```

---

> So I suggest to remove this variable as it is confusing.

Alright, i have removed that variable (and added an independent check that it has been assigned).

---

My example above still does not work.

---

I got your point, see the last commit where i swapped the argument checks and simplified the logic. OTOH, the assigned names for the input and noise in `@system(x⁺ = Ax + Bu + Du, input: u)` are `u` and `w` inside this function, because at that site it doesn't know what type of system this is. In other words, i don't think that we can handle this case here without bigger changes.

---

Ì think we could count how many time the `state`, `input` or `noise` was "found" in `extract_sum` to address this problem without bigger changes.

---

Something like this should work
```julia
function extract_sum(summands, state::Symbol, input::Symbol, noise::Symbol)
    params = Tuple{Any,Symbol}[]
    state_dim = 1
    got_state_dim = false

    num_state_assignements = 0
    num_input_assignements = 0
    num_noise_assignements = 0
    num_const_assignements = 0

    for summand in summands
        if @capture(summand, array_ * var_)
            if state == var
                push!(params, (array, :A))
                # obtain "state_dim" for later using in IdentityMultiple
                state_dim =  Expr(:call, :size, :($array), 1)
                got_state_dim = true
                num_state_assignements += 1

            elseif input == var
                push!(params, (array, :B))
                num_input_assignements += 1

            elseif noise == var
                push!(params, (array, :D))
                num_noise_assignements += 1

            else
                throw(ArgumentError("in the dynamic equation, the expression "*
                "$summand does not contain the state $state, the input $input "*
                "or the noise term $noise"))
            end
        elseif @capture(summand, array_)
            if got_state_dim
                identity = :(I($state_dim))
            else
                identity = 1.0
            end
            # if array == variable: field value equals identity
            if state == array
                push!(params, (identity, :A))
                num_state_assignements += 1
            elseif input == array
                push!(params, (identity, :B))
                num_input_assignements += 1
            elseif noise == array
                push!(params, (identity, :D))
                num_noise_assignements += 1
            else
                push!(params, (array, :c))
                num_const_assignements += 1
            end
        end
    end
    num_const_assignements > 1 && throw(ArgumentError("there is more than one constant term"))
    num_state_assignements > 1 && throw(ArgumentError("there is more than one state term $state"))
    num_input_assignements > 1 && throw(ArgumentError("there is more than one input term $input"))
    num_noise_assignements > 1 && throw(ArgumentError("there is more than one noise term $noise"))

    return params
end
```

---

> i don't think that we can handle this case here without bigger changes.

I think it is pretty simple: just compare the default values to the other values, or simpler: compare the final values *after* assigning the default values again.

---

for `@system(x⁺ = Ax + Bu + Du, input: u)`, [parse_system](https://github.com/JuliaReach/MathematicalSystems.jl/blob/a1d94d6658d6e7e6983e9fd5ef7783c20fb72b5f/src/macros.jl#L246) returns with `state_var -> :x`, `input_var -> :u` and `noise_var -> :w`, so there is no collision after assigning default values.

---

note that `parse_system` only changes defaults if it captures the statement `@capture(ex, (input = u_) | (u_ = input))`. There is also some logic to capture the [input](https://github.com/JuliaReach/MathematicalSystems.jl/blob/a1d94d6658d6e7e6983e9fd5ef7783c20fb72b5f/src/macros.jl#L273) from the form of the system, but it is not done for the noise.

---

> for `@system(x⁺ = Ax + Bu + Du, input: u)`, [parse_system](https://github.com/JuliaReach/MathematicalSystems.jl/blob/a1d94d6658d6e7e6983e9fd5ef7783c20fb72b5f/src/macros.jl#L246) returns with `state_var -> :x`, `input_var -> :u` and `noise_var -> :w`, so there is no collision after assigning default values.

I see. Still, from a programming point of view, I find it uncomfortable to assign a default value without checking compatibility with the other values that already exist. Here is another example where I added
```julia
    show(state_var)
    println()
    show(input_var)
    println()
    show(noise_var)
    println()
```
before and after the assignment with default values:
```julia
julia> @system(x⁺ = Ax + Bu + Du, noise: u)
# before assignment
:x
nothing
:u
# after assignment
:x
:u
:u
```

---

I get your point. But we "allow for conflicting variable names" in order to keep the interface as minimal as possible, as long as a variable is not used twice in the right-hand side.
`@system(x' = Ax + Bw, input:w)` would not work if we have more rigorous checks.

To see how we could throw an error for `@system(x⁺ = Ax + Bu + Du`
see https://github.com/JuliaReach/MathematicalSystems.jl/pull/164#discussion_r380758870

---

We can also make the default assignment in another function, i mean after `parse_system`, i.e. such that `parse_system` returns `nothing` if the variable hasn't been identifies at that point of the program.

---

>  `@system(x' = Ax + Bw, input:w)` would not work if we have more rigorous checks.

Only if you assign the values blindly. Of course you should not assign `w` as noise variable if there is no noise.

> But we "allow for conflicting variable names" in order to keep the interface as minimal as possible, as long as a variable is not used twice in the right-hand side.

But in my examples the variables were used twice.

> We can also make the default assignment in another function, i mean after `parse_system`, i.e. such that `parse_system` returns `nothing` if the variable hasn't been identified at that point of the program.

Okay, but I do not see how that solves the problem cases I mentioned.

---

> But in my examples the variables were used twice.

For this to work, we need to add https://github.com/JuliaReach/MathematicalSystems.jl/pull/164#discussion_r380758870 or something similar

---

Okay, I give up since nobody seems to care that two variables have the same value, which I thought was the whole point of this PR.